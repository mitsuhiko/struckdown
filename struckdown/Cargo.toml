[package]
name = "struckdown"
version = "0.1.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
edition = "2018"
license = "Apache-2.0"
description = "A structured markdown / commonmark library for Rust"
homepage = "https://github.com/mitsuhiko/struckdown"
repository = "https://github.com/mitsuhiko/struckdown"
keywords = ["cmark", "commonmark", "markdown", "rst"]
readme = "README.md"

[features]
default = ["external-processor", "syntect-processor", "html-sanitizer-processor", "autoanchors-processor"]
external-processor = ["tokio"]
syntect-processor = ["syntect"]
html-sanitizer-processor = ["ammonia", "uuid"]
autoanchors-processor = ["slug"]

[dependencies]
pulldown-cmark = "0.8.0"
serde = { version = "1.0.118", features = ["derive"] }
regex = "1.4.2"
lazy_static = "1.4.0"
v_htmlescape = "0.12.0"
serde_yaml = "0.8.14"
slug = { version = "0.1.4", optional = true }
serde_json = { version = "1.0.60", features = ["preserve_order"] }
tokio = { version = "1.0.0", features = ["rt", "process", "macros", "io-util"], optional = true }
syntect = { version = "4.5.0", optional = true }
ammonia = { version = "3.1.0", optional = true }
uuid = { version = "0.8.1", features = ["v4"], optional = true }
either = "1.6.1"

[dev-dependencies]
insta = { version = "1.3.0", features = ["glob"] }
