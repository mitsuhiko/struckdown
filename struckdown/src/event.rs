//! Implements all types emitted from an event.
use std::borrow::Cow;
use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::convert::TryInto;
use std::fmt::{self, Debug, Display};

use pulldown_cmark as cm;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

use crate::value::Value;

/// An internal string type.
///
/// This is not so much a string type as a container holding different
/// types of strings.  It does not even attempt to deref into a string
/// instead just provides a method [`Str::as_str`] to return the inner
/// value as string slice.
///
/// This string can be constructed from various other string types
/// via the [`From::from`] method.
#[derive(PartialEq, Eq, Hash, Clone)]
pub struct Str<'data> {
    inner: cm::CowStr<'data>,
}

impl<'data> Serialize for Str<'data> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(self.as_str())
    }
}

impl<'data, 'de> Deserialize<'de> for Str<'data> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let cow = Cow::<'data, str>::deserialize(deserializer)?;
        Ok(cow.into())
    }
}

impl From<String> for Str<'static> {
    fn from(value: String) -> Str<'static> {
        Str {
            inner: value.into(),
        }
    }
}

impl From<Box<str>> for Str<'static> {
    fn from(value: Box<str>) -> Str<'static> {
        Str {
            inner: cm::CowStr::Boxed(value),
        }
    }
}

impl<'data> From<&'data str> for Str<'data> {
    fn from(value: &'data str) -> Str<'data> {
        Cow::Borrowed(value).into()
    }
}

impl<'data> From<Cow<'data, str>> for Str<'data> {
    fn from(value: Cow<'data, str>) -> Str<'data> {
        match value {
            Cow::Borrowed(val) => Str {
                inner: cm::CowStr::Borrowed(val),
            },
            Cow::Owned(val) => Str {
                inner: cm::CowStr::Boxed(val.into()),
            },
        }
    }
}

impl<'data> Str<'data> {
    /// Creates a new string from a static literal.
    pub const fn new(val: &'data str) -> Str<'data> {
        Str {
            inner: cm::CowStr::Borrowed(val),
        }
    }

    /// Returns the contained string as string slice.
    pub fn as_str(&self) -> &str {
        &self.inner
    }

    /// Creates a string from a cmark string.
    pub(crate) fn from_cm_str(value: cm::CowStr<'data>) -> Str<'data> {
        Str { inner: value }
    }

    /// Slices the string down.
    pub(crate) fn slice(&self, start: usize, end: usize) -> Str<'data> {
        Str {
            inner: match self.inner {
                cm::CowStr::Borrowed(val) => cm::CowStr::Borrowed(&val[start..end]),
                cm::CowStr::Boxed(ref val) => {
                    cm::CowStr::Boxed(val[start..end].to_string().into_boxed_str())
                }
                cm::CowStr::Inlined(ref val) => {
                    cm::CowStr::Inlined((&val)[start..end].try_into().unwrap())
                }
            },
        }
    }

    /// Converts the string into a static version.
    fn into_static(self) -> Str<'static> {
        Str {
            inner: match self.inner {
                cm::CowStr::Borrowed(val) => cm::CowStr::Boxed(val.to_string().into_boxed_str()),
                cm::CowStr::Boxed(ref val) => cm::CowStr::Boxed(val.to_string().into_boxed_str()),
                cm::CowStr::Inlined(val) => cm::CowStr::Inlined(val),
            },
        }
    }
}

impl<'data> Display for Str<'data> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self.as_str(), f)
    }
}

impl<'data> Debug for Str<'data> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self.as_str(), f)
    }
}

impl<'data> PartialOrd for Str<'data> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.as_str().partial_cmp(other.as_str())
    }
}

impl<'data> Ord for Str<'data> {
    fn cmp(&self, other: &Self) -> Ordering {
        self.as_str().cmp(other.as_str())
    }
}

/// Location information for an annotated event.
#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
pub struct Location {
    /// Byte offset within the source document.
    pub offset: usize,
    /// Length in bytes in the source document.
    pub len: usize,
    /// Line in the source document (1 indexed).
    pub line: usize,
    /// Column in the source document (0 indexed).
    pub column: usize,
}

/// Event with annotations.
///
/// An annotated event is generally the same as an [`Event`] but it contains
/// optional annotations.  Annotations are generally just the location information
/// about where the event ocurred in the original source document.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnnotatedEvent<'data> {
    /// The actual event.
    #[serde(flatten)]
    pub event: Event<'data>,
    /// The optional location.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<Location>,
}

impl<'data> AnnotatedEvent<'data> {
    /// Shortcut to create annotated events.
    pub fn new<I: Into<Event<'data>>>(
        value: I,
        location: Option<Location>,
    ) -> AnnotatedEvent<'data> {
        AnnotatedEvent {
            event: value.into(),
            location,
        }
    }

    /// Converts the event into a static one.
    pub fn into_static(self) -> AnnotatedEvent<'static> {
        AnnotatedEvent {
            event: self.event.into_static(),
            location: self.location,
        }
    }
}

impl<'data, T: Into<Event<'data>>> From<T> for AnnotatedEvent<'data> {
    fn from(value: T) -> AnnotatedEvent<'data> {
        AnnotatedEvent::new(value, None)
    }
}

/// Alignment information.
#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Alignment {
    /// Undefined alignment
    None,
    /// Left aligned
    Left,
    /// Centered
    Center,
    /// Right aligned
    Right,
}

impl Default for Alignment {
    fn default() -> Alignment {
        Alignment::None
    }
}

/// Tag type
#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Tag {
    /// `<p>` tag equivalent.
    Paragraph,
    /// `<h1>` tag equivalent.
    Heading1,
    /// `<h2>` tag equivalent.
    Heading2,
    /// `<h3>` tag equivalent.
    Heading3,
    /// `<h4>` tag equivalent.
    Heading4,
    /// `<h5>` tag equivalent.
    Heading5,
    /// `<h6>` tag equivalent.
    Heading6,
    /// `<blockquote>` tag equivalent.
    BlockQuote,
    /// `<ol>` equivalent.
    OrderedList,
    /// `<ul>` equivalent.
    UnorderedList,
    /// `<li>` equivalent.
    ListItem,
    /// Defines a footnote.
    FootnoteDefinition,
    /// `<table>` equivalent.
    Table,
    /// `<thead>` equivalent.
    TableHeader,
    /// `<tbody>` equivalent.
    TableBody,
    /// `<tr>` equivalent.
    TableRow,
    /// `<th>` equivalent.
    TableHead,
    /// <td>` equivalent.
    TableCell,
    /// `<em>` equivalent.
    Emphasis,
    /// Alternative emphasis (might be underline).
    EmphasisAlt,
    /// `<strong>` equivalent.
    Strong,
    /// `<ss>` equivalent.
    Strikethrough,
    /// `<a>` equivalent.
    Link,
    /// `<div>` equivalent. Not used in syntax.
    Container,
    /// `<span>` equivalent. Not used in syntax.
    Span,
}

impl Tag {
    /// Returns the header level for the tag if the tag represents a header.
    ///
    /// This is useful because headers are represented by multiple
    /// distinct tags and this avoids having to match on multiple
    /// enum values.
    pub fn header_level(self) -> Option<usize> {
        match self {
            Tag::Heading1 => Some(1),
            Tag::Heading2 => Some(2),
            Tag::Heading3 => Some(3),
            Tag::Heading4 => Some(4),
            Tag::Heading5 => Some(5),
            Tag::Heading6 => Some(6),
            _ => None,
        }
    }

    /// Creates a start tag event.
    pub fn start_tag(self, attrs: Attrs<'_>) -> StartTagEvent<'_> {
        StartTagEvent { tag: self, attrs }
    }

    /// Creates an end tag event.
    pub fn end_tag(self) -> EndTagEvent {
        EndTagEvent { tag: self }
    }
}

fn is_default<T: Default + PartialEq>(v: &T) -> bool {
    *v == T::default()
}

/// Attributes for tags.
#[derive(Debug, Default, Serialize, Deserialize, Clone)]
pub struct Attrs<'data> {
    /// Holds the start for a list.  
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start: Option<u32>,
    /// Alignment information
    #[serde(default, skip_serializing_if = "is_default")]
    pub alignment: Alignment,
    /// An optional id for elements supporting it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Str<'data>>,
    /// Whitespace separated list of classes to attach to the element.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub class: Option<Str<'data>>,
    /// An optional title for elements supporting it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<Str<'data>>,
    /// Link reference target.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target: Option<Str<'data>>,
    /// Custom attributes.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom: Option<BTreeMap<Cow<'static, str>, Str<'data>>>,
}

impl<'data> Attrs<'data> {
    /// Returns `true` if all attrs are empty.
    pub fn is_empty(&self) -> bool {
        self.start.is_none()
            && self.alignment == Alignment::None
            && self.id.is_none()
            && self.class.is_none()
            && self.title.is_none()
            && self.target.is_none()
            && self.custom.is_none()
    }

    /// Converts the attrs into a static one.
    fn into_static(self) -> Attrs<'static> {
        Attrs {
            start: self.start,
            alignment: self.alignment,
            id: self.id.map(|x| x.into_static()),
            class: self.class.map(|x| x.into_static()),
            title: self.title.map(|x| x.into_static()),
            target: self.target.map(|x| x.into_static()),
            custom: self.custom.map(|custom| {
                custom
                    .into_iter()
                    .map(|(k, v)| (Cow::Owned(k.into_owned()), v.into_static()))
                    .collect()
            }),
        }
    }
}

/// Emitted at the start of a document.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DocumentStartEvent {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub front_matter: Option<Value>,
}

/// Emitted when a tag starts.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StartTagEvent<'data> {
    /// The tag that was started by this event.
    pub tag: Tag,
    /// Attached attributes to this event.
    #[serde(default, skip_serializing_if = "Attrs::is_empty")]
    pub attrs: Attrs<'data>,
}

/// Emitted when a tag ends.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct EndTagEvent {
    /// The tag that ends.
    pub tag: Tag,
}
/// A text event
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TextEvent<'data> {
    pub text: Str<'data>,
}

/// Not yet filled in interpreted text.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InterpretedTextEvent<'data> {
    /// The name of the role.
    pub role: Str<'data>,
    /// Text to the interpreted with that role.
    pub text: Str<'data>,
}

/// Code block
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CodeBlockEvent<'data> {
    /// The language argument to the code block.
    pub language: Option<Str<'data>>,
    /// Arguments to the code block.
    pub args: Option<BTreeMap<Str<'data>, Str<'data>>>,
    /// The raw code to be emitted.
    pub code: Str<'data>,
}

/// Directive block
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DirectiveEvent<'data> {
    /// The role of the directive.
    pub name: Str<'data>,
    /// The optional argument of the directive.
    pub argument: Option<Str<'data>>,
    /// The front matter if available.
    pub front_matter: Option<Value>,
    /// The directive body.
    pub body: Str<'data>,
}

/// Inline code
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InlineCodeEvent<'data> {
    /// The raw code to be emitted.
    pub code: Str<'data>,
}

/// An embedded image
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ImageEvent<'data> {
    /// The target location of the image
    pub target: Str<'data>,
    /// The optional alt text of the image
    pub alt: Option<Str<'data>>,
    /// The optional title of the image
    pub title: Option<Str<'data>>,
}

/// Embedded raw HTML
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RawHtmlEvent<'data> {
    pub html: Str<'data>,
}

/// A checkbox from a task list.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CheckboxEvent {
    pub checked: bool,
}

/// A reference to a footnote.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FootnoteReferenceEvent<'data> {
    pub target: Str<'data>,
}

/// An inline meta data event.
///
/// Meta data can be emitted during processing and can be collected by a
/// processor or final renderer to do something with it.  It's not defined
/// what the keys are or what happens if multiple values are emitted for
/// the same key.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct MetaDataEvent<'data> {
    pub key: Str<'data>,
    pub value: Value,
}

/// An event representing an error during processing
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ErrorEvent<'data> {
    pub title: Str<'data>,
    pub description: Option<Str<'data>>,
}

/// A event in a struckdown stream.
///
/// Struckdown events are not complete reflections of a markdown document.  In
/// fact it's not a goal to be able to go back from a struckdown event stream to
/// the original markdown document.  Certain normalizations are performed as part
/// of the parsing process to make it easier to perform stream processing.  As an
/// example images are no longer represented as tags but events.  Another example
/// are code tags which are normalized into an abstract code tag no matter the
/// original syntax.
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case", tag = "type")]
pub enum Event<'data> {
    DocumentStart(DocumentStartEvent),
    StartTag(StartTagEvent<'data>),
    EndTag(EndTagEvent),
    Text(TextEvent<'data>),
    InterpretedText(InterpretedTextEvent<'data>),
    CodeBlock(CodeBlockEvent<'data>),
    Directive(DirectiveEvent<'data>),
    InlineCode(InlineCodeEvent<'data>),
    Image(ImageEvent<'data>),
    RawHtml(RawHtmlEvent<'data>),
    SoftBreak,
    HardBreak,
    Rule,
    Checkbox(CheckboxEvent),
    FootnoteReference(FootnoteReferenceEvent<'data>),
    MetaData(MetaDataEvent<'data>),
    Error(ErrorEvent<'data>),
}

macro_rules! impl_from_event_type {
    ($variant:ident, $name:ty, $lifetime:tt) => {
        impl<$lifetime> From<$name> for Event<$lifetime> {
            fn from(value: $name) -> Self {
                Event::$variant(value)
            }
        }
    };
    ($variant:ident, $name:ty) => {
        impl From<$name> for Event<'static> {
            fn from(value: $name) -> Self {
                Event::$variant(value)
            }
        }
    };
}

impl_from_event_type!(DocumentStart, DocumentStartEvent);
impl_from_event_type!(StartTag, StartTagEvent<'data>, 'data);
impl_from_event_type!(EndTag, EndTagEvent);
impl_from_event_type!(Text, TextEvent<'data>, 'data);
impl_from_event_type!(InterpretedText, InterpretedTextEvent<'data>, 'data);
impl_from_event_type!(CodeBlock, CodeBlockEvent<'data>, 'data);
impl_from_event_type!(Directive, DirectiveEvent<'data>, 'data);
impl_from_event_type!(InlineCode, InlineCodeEvent<'data>, 'data);
impl_from_event_type!(Image, ImageEvent<'data>, 'data);
impl_from_event_type!(RawHtml, RawHtmlEvent<'data>, 'data);
impl_from_event_type!(Checkbox, CheckboxEvent);
impl_from_event_type!(FootnoteReference, FootnoteReferenceEvent<'data>, 'data);
impl_from_event_type!(MetaData, MetaDataEvent<'data>, 'data);
impl_from_event_type!(Error, ErrorEvent<'data>, 'data);

impl<'data> Event<'data> {
    /// Returns the contents as raw text.
    pub fn raw_text(&self) -> Option<&Str<'data>> {
        static NEWLINE: Str<'static> = Str::new("\n");
        static DOUBLE_NEWLINE: Str<'static> = Str::new("\n\n");
        match *self {
            Event::Text(TextEvent { ref text })
            | Event::InterpretedText(InterpretedTextEvent { ref text, .. })
            | Event::InlineCode(InlineCodeEvent { code: ref text, .. })
            | Event::CodeBlock(CodeBlockEvent { code: ref text, .. }) => Some(text),
            Event::Directive(DirectiveEvent { body: ref text, .. }) => Some(text),
            Event::SoftBreak => Some(&NEWLINE),
            Event::HardBreak => Some(&DOUBLE_NEWLINE),
            Event::Image(ImageEvent { ref alt, .. }) => alt.as_ref(),
            _ => None,
        }
    }

    /// Converts the event into a static one.
    pub fn into_static(self) -> Event<'static> {
        match self {
            Event::DocumentStart(value) => Event::DocumentStart(value),
            Event::StartTag(value) => Event::StartTag(StartTagEvent {
                tag: value.tag,
                attrs: value.attrs.into_static(),
            }),
            Event::EndTag(value) => Event::EndTag(value),
            Event::Text(value) => Event::Text(TextEvent {
                text: value.text.into_static(),
            }),
            Event::InterpretedText(value) => Event::InterpretedText(InterpretedTextEvent {
                role: value.role.into_static(),
                text: value.text.into_static(),
            }),
            Event::CodeBlock(value) => Event::CodeBlock(CodeBlockEvent {
                language: value.language.map(|x| x.into_static()),
                args: value.args.map(|x| {
                    x.into_iter()
                        .map(|(k, v)| (k.into_static(), v.into_static()))
                        .collect()
                }),
                code: value.code.into_static(),
            }),
            Event::Directive(value) => Event::Directive(DirectiveEvent {
                name: value.name.into_static(),
                argument: value.argument.map(|x| x.into_static()),
                front_matter: value.front_matter,
                body: value.body.into_static(),
            }),
            Event::InlineCode(value) => Event::InlineCode(InlineCodeEvent {
                code: value.code.into_static(),
            }),
            Event::Image(value) => Event::Image(ImageEvent {
                target: value.target.into_static(),
                alt: value.alt.map(|x| x.into_static()),
                title: value.title.map(|x| x.into_static()),
            }),
            Event::RawHtml(value) => Event::RawHtml(RawHtmlEvent {
                html: value.html.into_static(),
            }),
            Event::SoftBreak => Event::SoftBreak,
            Event::HardBreak => Event::HardBreak,
            Event::Rule => Event::Rule,
            Event::Checkbox(value) => Event::Checkbox(value),
            Event::FootnoteReference(value) => Event::FootnoteReference(FootnoteReferenceEvent {
                target: value.target.into_static(),
            }),
            Event::MetaData(value) => Event::MetaData(MetaDataEvent {
                key: value.key.into_static(),
                value: value.value.clone(),
            }),
            Event::Error(value) => Event::Error(ErrorEvent {
                title: value.title.into_static(),
                description: value.description.map(|x| x.into_static()),
            }),
        }
    }
}
