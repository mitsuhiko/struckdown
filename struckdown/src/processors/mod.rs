//! Stream processors for struckdown.
//!
//! This module provides stream processors that manipulate a struckdown event
//! stream to enhance it.  For instance a stream processor can automatically
//! add anchors to headers if they did not already get a header set by other
//! means.
#[macro_use]
mod utils;

mod toc;

#[cfg(feature = "autoanchors-processor")]
mod autoanchors;

#[cfg(feature = "external-processor")]
mod external;

#[cfg(feature = "syntect-processor")]
mod syntect;

#[cfg(feature = "html-sanitizer-processor")]
mod html_sanitizer;

use serde::Deserialize;

use crate::event::AnnotatedEvent;

pub use self::toc::{TableOfContents, TableOfContentsIter};

#[cfg(feature = "autoanchors-processor")]
pub use self::autoanchors::{AutoAnchors, AutoAnchorsIter};

#[cfg(feature = "external-processor")]
pub use self::external::{External, ExternalIter};

#[cfg(feature = "syntect-processor")]
pub use self::syntect::{Syntect, SyntectIter};

#[cfg(feature = "html-sanitizer-processor")]
pub use self::html_sanitizer::{HtmlSanitizer, HtmlSanitizerIter};

/// Common trait for all stream processors.
pub trait Processor {
    /// Applies the processor to an event stream.
    ///
    /// This consumes the processor.
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>;

    /// Applies the processor to an event stream.
    ///
    /// This attaches the processor by reference.
    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>;
}

macro_rules! builtin_processors {
    (
        $($(#[$attr:meta])* type $name:ident;)*
    ) => {
        /// Utility struct for processor configurations.
        #[derive(Debug, Deserialize, Clone)]
        #[serde(tag = "processor", rename_all = "snake_case")]
        pub enum BuiltinProcessor {
            $($(#[$attr])* $name(Box<$name>),)*
        }

        impl Processor for BuiltinProcessor {
            fn apply<'data>(
                self: Box<Self>,
                iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
            ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
                match *self {
                    $($(#[$attr])* Self::$name(options) => options.apply(iter),)*
                }
            }

            fn apply_ref<'data, 'options: 'data>(
                &'options self,
                iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
            ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
                match self {
                    $($(#[$attr])* Self::$name(options) => options.apply_ref(iter),)*
                }
            }
        }
    };
}

builtin_processors! {
    type TableOfContents;
    #[cfg(feature = "autoanchors-processor")]
    type AutoAnchors;
    #[cfg(feature = "external-processor")]
    type External;
    #[cfg(feature = "syntect-processor")]
    type Syntect;
    #[cfg(feature = "html-sanitizer-processor")]
    type HtmlSanitizer;
}
