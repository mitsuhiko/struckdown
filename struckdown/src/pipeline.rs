//! Abstracts event stream modifications.
use crate::event::{AnnotatedEvent, Event, Str};
use crate::parser::{Parser, ParserOptions};
use crate::processors::Processor;
use crate::value::Value;

/// Helper for applying preconfigured processors to an event stream.
pub struct Pipeline {
    parser: Parser,
    processors: Vec<Box<dyn Processor>>,
}

impl Default for Pipeline {
    fn default() -> Pipeline {
        Pipeline::new()
    }
}

impl Pipeline {
    /// Creates a new pipeline.
    pub fn new() -> Pipeline {
        Pipeline {
            parser: Parser::default(),
            processors: Vec::new(),
        }
    }

    /// Changes the parsing options.
    pub fn set_parser_options(&mut self, parser_options: &ParserOptions) {
        self.parser = Parser::new(parser_options);
    }

    /// Adds a processor to the pipeline
    pub fn add_processor<P: Processor + 'static>(&mut self, processor: P) {
        self.processors.push(Box::new(processor));
    }

    /// Applies the pipeline to a stream consuming the processor.
    pub fn apply<'data, I: Iterator<Item = AnnotatedEvent<'data>> + 'data>(
        self,
        iter: I,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        let mut iter = Box::new(iter) as Box<dyn Iterator<Item = AnnotatedEvent<'data>>>;
        for processor in self.processors {
            iter = processor.apply(iter);
        }
        iter
    }

    /// Applies the pipeline to a stream.
    pub fn apply_ref<'data, 'options: 'data, I: Iterator<Item = AnnotatedEvent<'data>> + 'data>(
        &'options self,
        iter: I,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        let mut iter = Box::new(iter) as Box<dyn Iterator<Item = AnnotatedEvent<'data>>>;
        for processor in &self.processors {
            iter = processor.apply_ref(iter);
        }
        iter
    }

    /// Parses and processes a document returning an event stream.
    pub fn process<'data, 'options: 'data>(
        &'options self,
        source: &'data str,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        self.apply_ref(Box::new(self.parser.parse(source)))
    }

    /// Extracts a meta data tree from an event stream.
    pub fn extract_metadata<
        'data: 'event,
        'options: 'data,
        'event,
        I: Iterator<Item = &'event AnnotatedEvent<'data>> + 'event,
    >(
        &'options self,
        iter: I,
    ) -> impl Iterator<Item = (&'event Str<'data>, &'event Value)> {
        iter.filter_map(|annotated_event| {
            if let Event::MetaData(ref md) = annotated_event.event {
                Some((&md.key, &md.value))
            } else {
                None
            }
        })
    }
}

#[test]
fn test_basic_pipeline() {
    use crate::html::to_html;
    use crate::processors::BuiltinProcessor;

    let mut pipeline = Pipeline::new();
    pipeline.add_processor(BuiltinProcessor::AutoAnchors(Default::default()));
    insta::assert_snapshot!(to_html(
        pipeline.process("# Hello World!\nAha"),
        &Default::default()
    ));
}
