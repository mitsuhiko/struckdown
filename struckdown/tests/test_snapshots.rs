use std::fs;

use struckdown::event::{AnnotatedEvent, DocumentStartEvent, Event};
use struckdown::html::to_html;
use struckdown::parser::parse;
use struckdown::pipeline::Pipeline;
use struckdown::processors::BuiltinProcessor;
use struckdown::value::Value;

use either::Either;

fn apply_configured_processors<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    processors: Vec<Value>,
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut pipeline = Pipeline::new();

    for processor in processors {
        pipeline.add_processor(serde_json::from_value::<BuiltinProcessor>(processor).unwrap());
    }

    pipeline.apply(Box::new(iter))
}

fn apply_processors<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut iter = iter.peekable();

    if let Some(document_start) = iter.peek() {
        if let Event::DocumentStart(DocumentStartEvent {
            front_matter: Some(ref front_matter),
        }) = document_start.event
        {
            if let Some(processors) = front_matter.get("processors").and_then(|x| x.as_array()) {
                return Either::Left(apply_configured_processors(processors.clone(), iter));
            }
        }
    }

    Either::Right(iter)
}

#[test]
fn test_parser() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let events: Vec<_> = apply_processors(parse(&source, &Default::default())).collect();
        insta::assert_yaml_snapshot!(events);
    });
}

#[test]
fn test_html() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let html = to_html(
            apply_processors(parse(&source, &Default::default())),
            &Default::default(),
        );
        insta::assert_snapshot!(html);
    });
}
