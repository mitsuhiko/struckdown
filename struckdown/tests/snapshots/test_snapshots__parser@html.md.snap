---
source: struckdown/tests/test_snapshots.rs
expression: events
input_file: struckdown/tests/inputs/html.md
---
- type: document_start
- type: start_tag
  tag: paragraph
  location:
    offset: 0
    len: 36
    line: 1
    column: 0
- type: text
  text: "This document has some inline html:"
  location:
    offset: 0
    len: 35
    line: 1
    column: 0
- type: end_tag
  tag: paragraph
  location:
    offset: 0
    len: 36
    line: 1
    column: 0
- type: raw_html
  html: "<div class=foo>\n"
  location:
    offset: 37
    len: 16
    line: 3
    column: 0
- type: raw_html
  html: "  bar\n"
  location:
    offset: 53
    len: 6
    line: 4
    column: 0
- type: raw_html
  html: "</div>\n"
  location:
    offset: 59
    len: 7
    line: 5
    column: 0
- type: start_tag
  tag: paragraph
  location:
    offset: 67
    len: 34
    line: 7
    column: 0
- type: text
  text: "And this has some mixed markdown:"
  location:
    offset: 67
    len: 33
    line: 7
    column: 0
- type: end_tag
  tag: paragraph
  location:
    offset: 67
    len: 34
    line: 7
    column: 0
- type: raw_html
  html: "<div class=foo>\n"
  location:
    offset: 102
    len: 16
    line: 9
    column: 0
- type: start_tag
  tag: paragraph
  location:
    offset: 119
    len: 16
    line: 11
    column: 0
- type: start_tag
  tag: emphasis
  location:
    offset: 119
    len: 15
    line: 11
    column: 0
- type: text
  text: markdown here
  location:
    offset: 120
    len: 13
    line: 11
    column: 1
- type: end_tag
  tag: emphasis
  location:
    offset: 119
    len: 15
    line: 11
    column: 0
- type: end_tag
  tag: paragraph
  location:
    offset: 119
    len: 16
    line: 11
    column: 0
- type: raw_html
  html: "</div>"
  location:
    offset: 136
    len: 6
    line: 13
    column: 0
